[package] turn MockZipReader into DirectoryReader and add test coverage

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59107 [package] turn MockZipReader into DirectoryReader and add test coverage**

Adding documentation, test coverage, and a missing method to the `DirectoryReader` class. `DirectoryReader` was previously named `MockZipReader`, and is used for operating on opened package archives via a `PackageImporter`.

Differential Revision: [D28760410](https://our.internmc.facebook.com/intern/diff/D28760410)

Commit:
[package] turn MockZipReader into DirectoryReader and add test coverage

[ghstack-poisoned]

Commit:
Update on "[package] turn MockZipReader into DirectoryReader and add test coverage"




[ghstack-poisoned]

Commit:
Update on "[package] turn MockZipReader into DirectoryReader and add test coverage"


Adding documentation, test coverage, and a missing method to the `DirectoryReader` class. `DirectoryReader` was previously named `MockZipReader`, and is used for operating on opened package archives via a `PackageImporter`. 

[ghstack-poisoned]

Lilyjjo commented on test/package/test_directory_reader.py:
```diff
@@ -0,0 +1,179 @@
+    "Tests that use temporary files are disabled in fbcode",
+)
+class DirectoryReaderTest(PackageTestCase):
+    """Tests use of DirectoryReader as accessor for opened packages."""
```
I created a main file to consolidate the tests for this class. I can instead put the relevant tests back into their normal sections. I wasn't sure which organization system was better

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28760410).

Commit:
Update on "[package] turn MockZipReader into DirectoryReader and add test coverage"


Adding documentation, test coverage, and a missing method to the `DirectoryReader` class. `DirectoryReader` was previously named `MockZipReader`, and is used for operating on opened package archives via a `PackageImporter`.

Differential Revision: [D28760410](https://our.internmc.facebook.com/intern/diff/D28760410)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28760410).

SplitInfinity commented on test/package/test_directory_reader.py:
```diff
@@ -0,0 +1,180 @@
+                    str(Path(temp_dir) / Path(filename).name)
+                )
+                dir_mod = dir_importer.load_pickle("res", "mod.pkl")
+
```
I think it would be useful to test that the `importlib.resources` API works with `DirectoryReader` as well. See the rest of the tests in `test/package/test_resources.py`.

SplitInfinity commented on torch/package/_directory_reader.py:
```diff
@@ -25,7 +25,16 @@ def storage(self):
-class MockZipReader(object):
+class DirectoryReader(object):
+    """
+    Class to allow PackageImporter to operate on unzipped packages. Methods
+    copy the behavior of the internal PyTorchFileReader class (which is used for
+    accessing packages in all other cases).
+
+    N.B.: ScriptObjects are not depickleable or accessible via this DirectoryReader
+    class due to ScriptObjects requiring an actual PyTorchFileReader instance.
```
Nit: This docstring is missing proper markup for class names and references, but it won't be rendered in the API reference anyway so maybe it is better the way it is now.

Lilyjjo replied:
Yeah, I'll add the proper markups if we make it public. I didn't want to bother with doing the setup needed to test to make sure the rendering was correct if we're not actually going to render this

SplitInfinity reviewed (commented):
I am not familiar with the full context on what this `DirectoryReader` is for, so I would take my review with a grain of salt.

I noticed some of the tests are very similar to existing tests (e.g. `DirectoryReaderTest.test_resource_reader` and `TestResources.test_resource_reader`). The only difference between them is how the packaged is stored and loaded. This suggests to me that for each test, we can add code (perhaps in `PackageTestCase`) that generates one version of the test that uses an `io.BytesIO` instance, another that uses a zip file, and another that uses an unzipped file (with each variant disabled as needed in fbcode). 

In the case of `TestResources.test_resource_reader`, the package setup code and assertions would remain the same, but the `PackageImporter` and `PackageExporter` instances would be provided by some magic test method that would return correctly configured instances based on whether the (generated) test uses a `io.BytesIO` instance, a zip file or an unzipped file.

I think that might be too much for this PR, but let me know what you think!

Commit:
Update on "[package] turn MockZipReader into DirectoryReader and add test coverage"


Adding documentation, test coverage, and a missing method to the `DirectoryReader` class. `DirectoryReader` was previously named `MockZipReader`, and is used for operating on opened package archives via a `PackageImporter`.

Differential Revision: [D28760410](https://our.internmc.facebook.com/intern/diff/D28760410)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28760410).

Lilyjjo commented:
> I noticed some of the tests are very similar to existing tests (e.g. DirectoryReaderTest.test_resource_reader and 
> TestResources.test_resource_reader). The only difference between them is ...

@SplitInfinity  I agree a larger refactoring of the tests would be nice! I don't really want to do that in this PR. And thank you for the thorough review, I addressed most of your comments

Lilyjjo commented on test/package/test_directory_reader.py:
```diff
@@ -0,0 +1,180 @@
+# -*- coding: utf-8 -*-
```
Yep!

Lilyjjo commented on test/package/test_directory_reader.py:
```diff
@@ -0,0 +1,180 @@
+                zip_file.extractall(path=temp_dir)
+                dir_importer = PackageImporter(
+                    str(Path(temp_dir) / Path(filename).name)
+                )
```
The temp file needs to stay open for the load_pickle call to work, I can switch the ordering of the with statements (but as long as the needed error is thrown I think it's fine?)

Lilyjjo commented on test/package/test_directory_reader.py:
```diff
@@ -0,0 +1,180 @@
+            importer = PackageImporter(str(Path(temp_dir) / Path(filename).name))
+            reader_one = importer.get_resource_reader("one")
+            with self.assertRaises(FileNotFoundError):
+                reader_one.resource_path("a.txt")
```
Good catch. I implemented the logic for this to work with the DirectoryReader pathway in PackageImporter 👍

Lilyjjo commented on test/package/test_directory_reader.py:
```diff
@@ -0,0 +1,180 @@
+            )
+            self.assertFalse(reader_one.is_resource("three"))
+            reader_one_contents = list(reader_one.contents())
+            self.assertSequenceEqual(
+                reader_one_contents, ["a.txt", "b.txt", "c.txt", "three"]
+            )
```
Nope, I added a 'sort' call to the list 👍

Lilyjjo commented on torch/package/package_importer.py:
```diff
@@ -624,6 +624,14 @@ def open_resource(self, resource):
+        ):
+            return os.path.join(
+                self.importer.zip_reader.directory, self.fullname, resource
+            )
```
I put this logic inside of this method because I didn't want to add it to the DirectoryReader itself. My thinking is that the DirectoryReader API should stay consistent with the PyTorchFileReader API, which doesn't have this functionality

Lilyjjo replied:
yep!

SplitInfinity reviewed (approved):
I think we should discuss some test refactoring at the next meeting but thanks for addressing my comments.

Commit:
Update on "[package] turn MockZipReader into DirectoryReader and add test coverage"


Adding documentation, test coverage, and a missing method to the `DirectoryReader` class. `DirectoryReader` was previously named `MockZipReader`, and is used for operating on opened package archives via a `PackageImporter`.

Differential Revision: [D28760410](https://our.internmc.facebook.com/intern/diff/D28760410)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28760410).

Commit:
Update on "[package] turn MockZipReader into DirectoryReader and add test coverage"


Adding documentation, test coverage, and a missing method to the `DirectoryReader` class. `DirectoryReader` was previously named `MockZipReader`, and is used for operating on opened package archives via a `PackageImporter`.

Differential Revision: [D28760410](https://our.internmc.facebook.com/intern/diff/D28760410)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28760410).
